File: day02/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

var numRegex = regexp.MustCompile(`\d+`)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	sum := 0
	sum2 := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		a, b, ok := strings.Cut(scanner.Text(), ": ")
		if !ok {
			log.Fatalln("Invalid line")
		}
		gameNum, err := strconv.Atoi(numRegex.FindString(a))
		if err != nil {
			log.Fatalln(err)
		}
		possible := true
		maxRed := 0
		maxGreen := 0
		maxBlue := 0
		rounds := strings.Split(b, "; ")
		for _, i := range rounds {
			cubes := strings.Split(i, ", ")
			for _, j := range cubes {
				a, b, ok := strings.Cut(j, " ")
				if !ok {
					log.Fatalln("Invalid cubes")
				}
				count, err := strconv.Atoi(a)
				if err != nil {
					log.Fatalln(err)
				}
				switch b {
				case "red":
					if count > 12 {
						possible = false
					}
					maxRed = max(maxRed, count)
				case "green":
					if count > 13 {
						possible = false
					}
					maxGreen = max(maxGreen, count)
				case "blue":
					if count > 14 {
						possible = false
					}
					maxBlue = max(maxBlue, count)
				default:
					log.Fatalln("Invalid color")
				}
			}
		}
		if possible {
			sum += gameNum
		}
		sum2 += maxRed * maxGreen * maxBlue
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	fmt.Println("Part 1:", sum)
	fmt.Println("Part 2:", sum2)
}

File: day02/src/main.rs
use once_cell::sync::Lazy;
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

static DIGIT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"\d+").unwrap());

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut sum1 = 0;
    let mut sum2 = 0;
    for line in reader.lines() {
        let mut possible = true;
        let mut max_red = 0;
        let mut max_green = 0;
        let mut max_blue = 0;
        let line = line?;
        let (a, b) = line.split_once(": ").ok_or("Invalid line")?;
        let game_num = DIGIT_REGEX
            .find(a)
            .and_then(|v| v.as_str().parse::<u32>().ok())
            .ok_or("Invalid game num")?;
        for v in b.split("; ").flat_map(|v| v.split(", ")) {
            let (a, b) = v.split_once(" ").ok_or("Invalid cubes")?;
            let count = a.parse::<u32>()?;
            match b {
                "red" => {
                    if count > 12 {
                        possible = false;
                    }
                    max_red = max_red.max(count);
                }
                "green" => {
                    if count > 13 {
                        possible = false;
                    }
                    max_green = max_green.max(count);
                }
                "blue" => {
                    if count > 14 {
                        possible = false;
                    }
                    max_blue = max_blue.max(count);
                }
                _ => Err("Invalid color")?,
            }
        }
        if possible {
            sum1 += game_num;
        }
        sum2 += max_red * max_green * max_blue;
    }

    println!("Part 1: {}", sum1);
    println!("Part 2: {}", sum2);
    Ok(())
}
